open-ended-capstone/second_etl.py

```diff
@@ -0,0 +1,194 @@
+        if len(primary_keys) != 1:
+            raise Exception("Simulator requires exactly one primary key")
+        self._primary_key = primary_keys[0]
+        self._update_columns = [col for col in columns if col.get_type() == 'VARCHAR']  # restrict to VARCHAR update
+        if len(primary_keys) == 0:
+            raise Exception("Need at least one VARCHAR for update")
+
+    def get_name(self):
+
+        return self._name
+
+    def get_columns(self):
+
+        return self._columns
+
+    def get_primary_key(self):
+
+        return self._primary_key
+
+    def get_column_names(self):
+
+        return [col.get_name() for col in self._columns] 
+
+    def get_update_column(self):
+
+        i = random.randint(0,len(self._update_columns)-1)   
+        return self._update_columns[i]
+
+PRODUCT_CREATE_SQL_PG = \
+" CREATE TABLE IF NOT EXISTS Product ("\
+" product_id INTEGER NOT NULL,  "\
+" product_name VARCHAR(80) NOT NULL,"\
+" product_description VARCHAR(255) NULL DEFAULT NULL,"\
+" product_category VARCHAR(80) NULL DEFAULT NULL,"\
+" product_brand VARCHAR(80) NULL DEFAULT NULL,"\
+" product_preferred_supplier_id INTEGER NULL DEFAULT NULL,"\
+" product_dimension_length FLOAT(11) NULL DEFAULT NULL,"\
+" product_dimension_width FLOAT(11) NULL DEFAULT NULL,"\
+" product_dimension_height FLOAT(11) NULL DEFAULT NULL,"\
+" product_introduced_date DATE NULL DEFAULT NULL,"\
+" product_discontinued BOOLEAN NULL DEFAULT FALSE,"\
+" product_no_longer_offered BOOLEAN NULL DEFAULT FALSE,"\
+" PRIMARY KEY (product_id));" 
+
+PRODUCT_TABLE = Table('product', 
+                Column('product_id', 'INTEGER', isPrimaryKey=True),
+                Column('product_name', 'VARCHAR'),
+                Column('product_description', 'VARCHAR'),
+                Column('product_category', 'VARCHAR'),
+                Column('product_brand', 'VARCHAR'),
+                Column('product_preferred_supplier_id', 'INTEGER'),
+                Column('product_dimension_length', 'FLOAT'),
+                Column('product_dimension_width', 'FLOAT'),
+                Column('product_dimension_height', 'FLOAT'),
+                Column('product_introduced_date', 'DATE'),
+                Column('product_discontinued', 'BOOLEAN'),
+                Column('product_no_longer_offered', 'BOOLEAN'))
+
+DEFAULT_INSERT_VALUES = { 'INTEGER' : 98,
+                          'VARCHAR' : 'AAA',
+                          'FLOAT' : 5.0,
+                          'REAL'  : 5.0,
+                          'DATE'  : '2021-02-11 12:52:47',
+                          'TINYINT' : 0,
+                          'BOOLEAN' : True }
+
+def create_table_src(conn, create_sql):
+
+    cur = conn.cursor()
+    cur.execute(create_sql)
+    conn.commit()
+    cur.close()
+
+def create_table_trg(conn, table):
+    pass
+
+def extract_src_to_trg(source, target, table):
+    pass
+
+def incremental_load_src(src_conn, table, n_inserts, n_updates):
```

Comment:
not clear what types of objects are expected to be passed as variables. 

would recommend using types annotations